runner/src/runner.rs

```diff
@@ -359,6 +364,59 @@ async fn run_binary(runner: &mut impl RunnerExt, stdout_mode: impl Into<Stdio>)
     Ok(child)
 }
 
+async fn run(runner: &mut impl RunnerExt) -> Result<(), Error> {
+    runner.subscribe_to_shutdown_signals()?;
+    // Create all directories for the `download_path` if they don't already exist.
+    fs::create_dir_all(&runner.download_path())?;
+    let release = runner.try_get_release(false).await?.expect("No current release");
+    // WARNING: This will overwrite any pre-existing binary with the same name
+    // TODO: Check if a release with the same version is already at the `download_path`
+    runner.download_binary(release).await?;
+
+    runner.run_binary().await?;
+
+    let mut loop_delay_counter = Duration::from_secs(0);
+    loop {
+        if runner.shutdown_flag().load(Ordering::Relaxed) {
+            // Terminate child process
+            runner.terminate_proc_and_wait()?;
+
+            // Break the loop to gracefully terminate the main process.
+            break;
+        }
+
+        tokio::time::sleep(OS_SIGNAL_PROCESSING_DELAY).await;
+        loop_delay_counter = loop_delay_counter.add(OS_SIGNAL_PROCESSING_DELAY);
+
+        // Only check for a new release if `BLOCK_TIME` has elapsed since the last check
+        if loop_delay_counter.lt(&BLOCK_TIME) {
+            continue;
+        }
+        // Reset the delay counter
+        loop_delay_counter = Duration::from_secs(0);
+
+        if let Some(new_release) = runner.try_get_release(false).await? {
+            let maybe_downloaded_release = runner.downloaded_release();
+            let downloaded_release = maybe_downloaded_release.as_ref().ok_or(Error::NoDownloadedRelease)?;
+            if new_release.uri != downloaded_release.release.uri {
+                // Wait for child process to finish completely.
+                // To ensure there can't be two vault processes using the same Bitcoin wallet.
+                runner.terminate_proc_and_wait()?;
+
+                // Delete old release
+                runner.delete_downloaded_release()?;
+
+                // Download new release
+                runner.download_binary(new_release).await?;
+
+                // Run the downloaded release
+                runner.run_binary().await?;
+            }
+        }
+    }
+    Ok(())
+}
```

Comment:
What is your reason for converting this into a standalone function?

Reply:
So I could test the actual `run` function with a `MockRunner`. It's the same pattern I used for all other `RunnerExt` trait functions.